### PR TITLE
Added code to generate rhat.

### PR DESCRIPTION
**Pull request by simonleandergrimm**
Code output for `fit_panel.py`:

```
Model fitting of panel-amplified samples complete
R-hat statistics:
crits_christoph, SARS-COV-2: rhat=1.0033576194637268
rothman, SARS-COV-2: rhat=1.0016563699601826
crits_christoph, HIV: rhat=1.0163708593499354
rothman, HIV: rhat=1.033028485235259
crits_christoph, CMV: rhat=1.0036299779197166
rothman, CMV: rhat=1.0184810844050918
crits_christoph, HSV-1: rhat=1.0020378384750617
rothman, HSV-1: rhat=1.0079267949758453
crits_christoph, Influenza A: rhat=1.0093984339393791
rothman, Influenza A: rhat=1.0179386520665972
crits_christoph, Influenza B: rhat=1.002628817358168
rothman, Influenza B: rhat=1.0081940505926346
crits_christoph, Norovirus (GI): rhat=1.0044655208652444
rothman, Norovirus (GI): rhat=1.0019271661313076
crits_christoph, Norovirus (GII): rhat=1.0022181669527972
rothman, Norovirus (GII): rhat=1.0039390279503166
crits_christoph, HPV: rhat=1.0175493940016302
rothman, HPV: rhat=1.0584053439232826
crits_christoph, EBV: rhat=1.0041078984492227
rothman, EBV: rhat=1.1180052354055459
crits_christoph, HCV: rhat=1.0023509799019954
rothman, HCV: rhat=1.035203209771713
```

Code output for `fit.py`:
```
Model fitting for non-enriched samples complete
R-hat statistics:
crits_christoph, SARS-COV-2: rhat=1.0035478901591979
rothman, SARS-COV-2: rhat=1.0072266373370191
spurbeck, SARS-COV-2: rhat=1.0229288554487868
crits_christoph, HIV: rhat=1.008222312666826
rothman, HIV: rhat=1.051128872165018
spurbeck, HIV: rhat=1.018594029447667
brinch, HIV: rhat=1.2628348565111083
crits_christoph, CMV: rhat=1.0019712797357265
rothman, CMV: rhat=1.0580356377994862
spurbeck, CMV: rhat=1.0071872038657101
brinch, CMV: rhat=1.0768627213617208
crits_christoph, HSV-1: rhat=1.0106495893429668
rothman, HSV-1: rhat=1.0091339201462586
spurbeck, HSV-1: rhat=1.0152715520474924
brinch, HSV-1: rhat=1.018951965878695
crits_christoph, Influenza A: rhat=1.0256962411667108
rothman, Influenza A: rhat=1.0359093723533825
spurbeck, Influenza A: rhat=1.0266966837161098
crits_christoph, Influenza B: rhat=1.005996009329864
rothman, Influenza B: rhat=1.0096232404303642
spurbeck, Influenza B: rhat=1.0123081796124465
crits_christoph, Norovirus (GI): rhat=1.0012973209330207
rothman, Norovirus (GI): rhat=1.00012813580041
spurbeck, Norovirus (GI): rhat=1.0029339874864631
crits_christoph, Norovirus (GII): rhat=1.0038702292460073
rothman, Norovirus (GII): rhat=1.004824004135918
spurbeck, Norovirus (GII): rhat=1.001288130654227
crits_christoph, HPV: rhat=1.0018490240117366
rothman, HPV: rhat=1.0353004445176937
spurbeck, HPV: rhat=1.0102103276793883
brinch, HPV: rhat=1.018719940102656
crits_christoph, EBV: rhat=1.0043951853874542
rothman, EBV: rhat=1.036357706593936
spurbeck, EBV: rhat=1.0112248249002724
brinch, EBV: rhat=1.0191225816934109
crits_christoph, HCV: rhat=1.0025685018555885
rothman, HCV: rhat=1.0064320652863488
spurbeck, HCV: rhat=1.0354271624283782
brinch, HCV: rhat=1.1079212023633667
```

### CODE REVIEW

**Review by dp-rice — APPROVED** [review]
LGTM. A few suggestions:

- Since we're doing the same thing twice, it would probably be better to move the code to a method `model.get_rhat()` and call that in the various places
- I realize that there are a few mysterious things in my rhat code that could use comments. I'll suggest them inline.

File: fit.py [hunk 1]
```diff
@@ -67,6 +68,18 @@ def start(num_samples: int, plot: bool) -> None:
             )
             input_data.append(model.input_df.assign(**metadata))
             output_data.append(model.get_coefficients().assign(**metadata))
+
+            coeffs = model.get_coefficients()
+
+            rhat_data = (
+                coeffs[coeffs["location"] == "Overall"]["b"]
+                .to_numpy()
+                .reshape((num_samples, -1))
```
Comment (dp-rice) [inline fit.py:77]:
```suggestion
                # Un-stack the MCMC chains
                .reshape((num_samples, -1))
```

**Comment by simonleandergrimm** [comment]
Followed your suggestions. lmk if I can merge!

**Comment by dp-rice** [comment]
Two small things:

- I suggested one more comment above.
- I think that `Model` should have attributes `num_samples` and `num_chains` that are initially `None` and then get set when you call `fit_model`. Then when you can use `self.num_samples` in `get_rhat` rather than having to pass the value again (and potentially giving it the wrong value).

**Comment by simonleandergrimm** [comment]
`Model` has no attribute `num_samples`; it just gets defined as an integer when calling fit.py and is then used like an integer throughout.

I nevertheless did something in this direction, but I'm not sure how much clearer this is.

In `fit.py`
```
model.fit_model(num_samples=num_samples)
model.num_samples = num_samples
rhat = model.get_rhat()
```

In `stats.py`:
```
def get_rhat(self) -> float:
    num_samples = self.num_samples
    coeffs = self.get_coefficients()
    rhat_data = (
        coeffs[coeffs["location"] == "Overall"]["b"]
        .to_numpy()
        # Un-stack the MCMC chains
        .reshape((num_samples, -1))
        .T
    )
    rhat = az.rhat(rhat_data)
    return that
```

**Comment by simonleandergrimm** [comment]
> Two small things:
> 
> * I suggested one more comment above.

I might be missing something, I don't see another suggested inline comment.

File: fit.py [hunk 2]
```diff
@@ -67,6 +68,18 @@ def start(num_samples: int, plot: bool) -> None:
...
+                .T
```
[the remark this answers is not in the document]
Comment (dp-rice) [reply fit.py]: @simonleandergrimm the second one here

**Comment by dp-rice** [comment]
> > Two small things:
> > 
> > * I suggested one more comment above.
> 
> I might be missing something, I don't see another suggested inline comment.

I tagged you above

**Comment by simonleandergrimm** [comment]
> > `Model` has no attribute `num_samples`; it just gets defined as an integer when calling fit.py and is then used like an integer throughout.
> 
> I'm suggesting that you add one to the class definition.

I still don't see it, neither here or in my email. I just see the suggestion to add `# Un-stack the MCMC chains, ` which I've added.
image


Can you tell me what comment you wanted to add?

File: stats.py [hunk 3]
```diff
@@ -192,6 +193,18 @@ def fit_model(self, num_chains: int = 4, num_samples: int = 1000) -> None:
...
         self.output_df = self.fit.to_frame()
 
+    def get_rhat(self, num_samples: int) -> float:
+        coeffs = self.get_coefficients()
+        rhat_data = (
+            coeffs[coeffs["location"] == "Overall"]["b"]
+            .to_numpy()
+            # Un-stack the MCMC chains
+            .reshape((num_samples, -1))
+            .T
```
Comment (dp-rice) [inline stats.py:203]:
```suggestion
            # az.rhat() expects shape (num_chains, num_samples)
            .T
```

Comment (dp-rice) [reply stats.py:203]: @simonleandergrimm trying again.

**Comment by dp-rice** [comment]
I realized what the problem was. The suggestion was stuck as "pending" a review

File: stats.py (hunk 3 above)
Comment (simonleandergrimm) [reply stats.py:207]: added!

**Comment by simonleandergrimm** [comment]
Ah ok. I added the things you suggested, let me know what you think.

**Comment by dp-rice** [comment]
Looks good!

**Pull request merged**